emcc.py

```diff
@@ -3064,40 +3062,52 @@ def get_language_mode(args):
   language_mode = get_language_mode(newargs)
   use_cxx = 'c++' in language_mode or run_via_emxx
 
-  def get_clang_command(src_file):
-    return compiler + get_cflags(state.orig_args, use_cxx) + compile_args + [src_file]
+  def get_clang_command():
+    return compiler + get_cflags(state.orig_args, use_cxx) + compile_args
 
-  def get_clang_command_preprocessed(src_file):
-    return compiler + get_clang_flags(state.orig_args) + compile_args + [src_file]
+  def get_clang_command_preprocessed():
+    return compiler + get_clang_flags(state.orig_args) + compile_args
 
-  def get_clang_command_asm(src_file):
-    return compiler + get_target_flags() + compile_args + [src_file]
+  def get_clang_command_asm():
+    return compiler + get_target_flags() + compile_args
 
   # preprocessor-only (-E) support
   if state.mode == Mode.PREPROCESS_ONLY:
-    for input_file in [x[1] for x in input_files]:
-      cmd = get_clang_command(input_file)
-      if options.output_file:
-        cmd += ['-o', options.output_file]
-      # Do not compile, but just output the result from preprocessing stage or
-      # output the dependency rule. Warning: clang and gcc behave differently
-      # with -MF! (clang seems to not recognize it)
-      logger.debug(('just preprocessor ' if state.has_dash_E else 'just dependencies: ') + ' '.join(cmd))
-      shared.check_call(cmd)
+    inputs = [i[1] for i in input_files]
+    cmd = get_clang_command() + inputs
+    if options.output_file:
+      cmd += ['-o', options.output_file]
+    # Do not compile, but just output the result from preprocessing stage or
+    # output the dependency rule. Warning: clang and gcc behave differently
+    # with -MF! (clang seems to not recognize it)
+    logger.debug(('just preprocessor ' if state.has_dash_E else 'just dependencies: ') + ' '.join(cmd))
+    shared.check_call(cmd)
     return []
 
   # Precompiled headers support
   if state.mode == Mode.PCH:
-    headers = [header for _, header in input_files]
-    for header in headers:
+    inputs = [i[1] for i in input_files]
+    for header in inputs:
       if not shared.suffix(header) in HEADER_ENDINGS:
         exit_with_error(f'cannot mix precompiled headers with non-header inputs: {headers} : {header}')
-      cmd = get_clang_command(header)
-      if options.output_file:
-        cmd += ['-o', options.output_file]
-      logger.debug(f"running (for precompiled headers): {cmd[0]} {' '.join(cmd[1:])}")
-      shared.check_call(cmd)
-      return []
+    cmd = get_clang_command() + inputs
+    if options.output_file:
+      cmd += ['-o', options.output_file]
+    logger.debug(f"running (for precompiled headers): {cmd[0]} {' '.join(cmd[1:])}")
+    shared.check_call(cmd)
+    return []
+
+  if state.mode == Mode.COMPILE_ONLY:
```

Comment:
Why is this limited to compile-only mode? Wouldn't it work for compile-and-link too?

Reply:
It compile-and-link mode it more complicated because we write done want to leave any object files in the current working directory.   If we could find a way to extent this later to cover all cases that would be awesome but for now we want control over the output filenames in that case (just like the clang driver internally does).

Reply:
I thought we could just pass the args to Clang so e.g. `emcc foo.c bar.c -o output.js` becomes `clang foo.c bar.c -o output.wasm` but I guess it's not that simple?

Reply:
Yes, the problem is that we don't really on clang as out linker-driver, we do it outselves.  We run wam-ld and then a bunch of post-link tools from emcc.py. 

If we could convert to using clang as the linker driver (i.e. have clang run wasm-ld for us) that would a really good step in the right direction.

Reply:
In other words, when linking we currently manage the names and locations of the temporary object files, and not clang.   Maybe we could improve on this in the future.

Reply:
Makes sense. Leaving as a future improvement sgtm.